src/debugger/mock-ui/UISession.js

```diff
@@ -0,0 +1,212 @@
+import { DataHandler } from "./DataHandler.js";
+
+//separator for messages according to the protocol
+const TWO_CRLF = "\r\n\r\n";
+
+/* Represents one debugging session in the CLI.
+ * Read in user input from the command line, parses the input into commands,
+ * sends the commands to the adapter and process any responses
+*/
+export class UISession {
+  constructor(proc: Process, adapterPath: string, prepackCommand: string) {
+    this._proc = proc;
+    this._adapterPath = adapterPath;
+    this._prepackCommand = prepackCommand;
+    this._sequenceNum = 0;
+    this._invalidCount = 0;
+    this._dataHandler = new DataHandler();
+  }
+  // the parent (i.e. ui) process
+  _proc: Process;
+  //path to the debug adapter
+  _adapterPath: string;
+  // the child (i.e. adapter) process
+  _adapterProcess: child_process.ChildProcess;
+
+  // id number for each message sent
+  _sequenceNum: number;
+  // interface to read in input from the CLI client
+  _reader: readline.Interface;
+  // number of invalid commands
+  _invalidCount: number;
+  // command to start Prepack with
+  _prepackCommand: string;
+  // handler for any received messages
+  _dataHandler: DataHandler;
+
+  _startAdapter() {
+    let adapterArgs = [this._adapterPath, "--prepack", this._prepackCommand];
+    this._adapterProcess = child_process.spawn("node", adapterArgs);
+    this._proc.on("exit", () => {
+      this.shutdown();
+    });
+    this._proc.on("SIGINT", () => {
+      this.shutdown();
+    });
+    this._adapterProcess.on("exit", () => {
+      this.shutdown();
+    });
+    this._adapterProcess.stdout.on("data", (data: Buffer) => {
+      //handle the received data
+      this._dataHandler.handleData(data, this._processMessage.bind(this));
+      //ask the user for the next command
+      this._reader.question("(dbg) ", (input: string) => {
+        this._dispatch(input);
+      });
+    });
+    this._adapterProcess.stderr.on("data", (data: Buffer) => {
+      console.error(data.toString());
+      this.shutdown();
+    });
+  }
+
+  // called from data handler to process a received message
+  _processMessage(message: string): void {
+    try {
+      let msg = JSON.parse(message);
+      if (msg.type === "event") {
+        this._processEvent(msg);
+      } else if (msg.type === "response") {
+        this._processResponse(msg);
+      }
+    } catch (e) {
+      console.error(e);
+      console.error("Invalid message: " + message.slice(0, 1000));
+    }
+  }
+
+  _processEvent(event: DebugProtocol.Event) {
+    // to be implemented
+    console.log(event);
+  }
+
+  _processResponse(response: DebugProtocol.Response) {
+    // to be implemented
+    console.log(response);
+  }
+
+  // execute a command if it is valid
+  // returns whether the command was valid
+  _executeCommand(input: string): boolean {
+    let parts = input.split(" ");
+    let command = parts[0];
+
+    // for testing purposes, init and configDone are made into user commands
```

Comment:
Maybe pose this comment as a TODO

Reply:
For the CLI, we can keep these two commands as being from user input. This way we can control tests on them using the CLI. In an actual IDE, these will be done without user input.